text/0000-std-aware-cargo.md

```diff
@@ -0,0 +1,335 @@
+
+Right now these are built using the global lock file in `rust-lang/rust`. Should this always be true? How should Cargo handle this gracefully?
+
+## Should profile changes always prompt a rebuild of `core`/`std`?
+
+For example, if a user sets their debug build to use `opt-level = 'z'`, should this rebuild `core`/`std` to use that opt level? Or should an additional flag, such as `apply-to-sysroot` be required to opt-in to this behavior, unless otherwise needed?
+
+This could increase compile times for users that have set profile overrides, but have not previously needed a custom `core` or `std`.
+
+Another option in this area is to force the use of profile overrides, as specified by [RFC2822](https://github.com/rust-lang/rfcs/blob/master/text/2282-profile-dependencies.md).
+
+## Should providing a custom `core` or `std` require a nightly compiler?
+
+It is currently unknown whether it is possible to provide a custom version of `core` or `std` without unstable features, as there are some compiler intrinsics and "magic" that are necessary (the format macros and box keyword come to mind).
+
+I initially wrote the RFC in this manner, however I was later convinced this was not possible to do.
+
+I am of the opinion that if you could, then it should be allowed to use a stable compiler, but that might be too theoretical for this RFC.
+
+We could also move forward with the current restriction to nightly, and allow that to be lifted later by a follow-on RFC if this is possible and necessary.
+
+## Should we allow configurable `core` and `std`
+
+If we are to uphold stability guarantees for all configurations of `core` and `std`, this could require testing 2^(n+m) versions of Rust, where `n` is the number of `core` features, and `m` is the number of `std` features. This would have a negative impact on CI times.
+
+# Future possibilities
+[future-possibilities]: #future-possibilities
+
+## Unified `core` and `std`
+
+With the mechanisms specified above, it could be possible to remove the concept of `core` and `std` from the user, leaving only `std`.
+
+By using stable feature flags for `std`, we could say that `std` as a crate with `default-features = false` would essentially be `no_core`, or with `features = ["core"]`, we would be the same as `no_std`.
+
+This abstraction may not map to the actual implementation of `libcore` or `libstd`, but instead be an abstraction layer for the end developer.
+
+## Stop shipping pre-compiled `core` and `std`
+
+With the ability to build these crates on demand, we may want to decide not to ship `target` bundles for any users.
+
+This would come at a cost of increased compile times, at least for the first build, if the artifacts are cached globally. However it would remove a mental snag of having to sometimes run `rustup target add`, and confusion from some users why parts of `std` and `core` have different optimization settings (particularly for debug builds) when debugging.
```

Comment:
Wouldn’t this RFC as-is already make `rustup target add` unnecessary? (In at least some cases.) When the pre-compiled standard library is not available for a given target, Cargo falls back to compiling it. We could emit a warning for target that are known to have a pre-compiled copy available through rustup but not installed locally.